crates/wasm-mutate/src/mutators/codemotion/ir/mod.rs

```diff
@@ -0,0 +1,501 @@
+        blocks: Vec<usize>,
+    ) -> Self {
+        AST {
+            root,
+            nodes,
+            ifs,
+            loops,
+            blocks,
+        }
+    }
+
+    /// Returns true if the AST has if-else nodes
+    ///
+    pub fn has_if(&self) -> bool {
+        !self.ifs.is_empty()
+    }
+
+    /// Returns true if the AST has loop nodes
+    ///
+    pub fn has_loop(&self) -> bool {
+        !self.loops.is_empty()
+    }
+
+    /// Returns true if the AST has block nodes
+    ///
+    pub fn has_block(&self) -> bool {
+        !self.blocks.is_empty()
+    }
+
+    /// Returns the node indexes corresponding to if-else nodes
+    ///
+    pub fn get_ifs(&self) -> &[usize] {
+        &self.ifs
+    }
+
+    /// Returns the `Root` node index of the AST
+    ///
+    pub fn get_root(&self) -> usize {
+        self.root
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Node {
+    /// If-Else node (then branch node indexes, alternative branch node indexes, type of the IfElse node)
+    IfElse(Vec<usize>, Option<Vec<usize>>, TypeOrFuncType),
+    /// Code node (pseudo basic block) (range of the operators in the input function)
+    Code { range: Range },
+    /// Loop node (pseudo basic block) (sorted children node indexes, type of the Loop node)
+    Loop(Vec<usize>, TypeOrFuncType),
+    /// Block node (pseudo basic block) (sorted children node indexes, type of the Block node)
+    Block(Vec<usize>, TypeOrFuncType),
+    /// Special node to wrap the root nodes of the AST
+    Root(Vec<usize>),
+}
+
+impl ASTBuilder {
+    /// Returns an AST from the operators collected from the Wasm function
+    pub fn build_ast<'a>(&self, operators: &'a [OperatorAndByteOffset]) -> crate::Result<AST> {
+        let (root, nodes, ifs, loops, blocks) = self.parse(operators)?;
+        Ok(AST::new(root, nodes, ifs, loops, blocks))
+    }
+
+    /// Parsing algorith to construct the AST
+    fn parse<'a>(
+        &self,
+        operators: &'a [OperatorAndByteOffset],
+    ) -> crate::Result<(usize, Vec<Node>, Vec<usize>, Vec<usize>, Vec<usize>)> {
+        #[derive(Debug)]
+        enum State {
+            If,
+            Else,
+            Loop,
+            Block,
+            Root,
+        }
+
+        let nodes = Vec::new();
+
+        let mut ifs = Vec::new();
+        let mut loops = Vec::new();
+        let mut blocks = Vec::new();
+
+        let nodes = RefCell::new(nodes);
+
+        let mut frame_stack = vec![(State::Root, None)];
+        let state_stack = RefCell::new(Vec::new());
+        let current_parsing = RefCell::new(Vec::new());
+
+        let push_state = || {
+            state_stack.borrow_mut().push(current_parsing.clone());
+            *current_parsing.borrow_mut() = vec![]
+        };
+        let pop_state = || {
+            state_stack
+                .borrow_mut()
+                .pop()
+                .expect("Missing state")
+                .borrow()
+                .clone()
+        };
```

Comment:
I think we could remove these `RefCell`s and make this a little more idiomatic by passing in a mutable borrow of the stack state to these closures.

It may also make sense to define something like

```rust
struct ParseContext {
    current: ...,
    stack: Vec<...>,
    frames: Vec<...>,
}
```

and pass mutable references to this to the `push_state`/`pop_state` methods. Even better, `push_state` and `pop_state` could be `&mut self` methods of `ParseContext`.